### DISCUSSION TITLE
sum crc32

### DISCUSSION THREAD

**Discussion by tooptoop4** [dathere/qsv#2179] [Q&A] [answered]
help with https://github.com/jqnatividad/qsv/pull/2121#issuecomment-2373164657

**Comment by jqnatividad** [comment]
Hi @tooptoop4 ,

Can you run `qsv --version` and share it here?

What kind of machine are you using? How big is the file? How many rows? How many columns?  Can you share the output of `qsv stats --everything` as well? Did you index the file? How long did it take to run?

**Reply by tooptoop4** [reply]
@jqnatividad 0.135.0 version. debian python docker image, 800k rows, 150 cols. no index

**Reply by jqnatividad** [reply]
For reference, here's your existing pipeline:

```
/clis/qsv apply operations crc32 a_column --new-column bobb input.csv | /clis/qsv select bobb > cr.txt; /clis/qsv sqlp cr.txt 'select sum(bobb) from cr'
```

FYI, `sqlp` has a heuristic that uses polar's mem-mapped `read_csv` table function that only works if the extension is `.csv`. Otherwise, it loads the CSV using a slower approach, as in your case as you're using a `.txt` extension:

https://github.com/jqnatividad/qsv/blob/293891076d6af1025ea5d79a9e7bb677277c715e/src/cmd/sqlp.rs#L660-L712

Just change `cr.txt` to `cr.csv`:
```
/clis/qsv apply operations crc32 a_column --new-column bobb input.csv | /clis/qsv select bobb > cr.csv; /clis/qsv sqlp cr.csv 'select sum(bobb) from cr'
```

You should see a big performance difference between the two versions. Can you share the delta here?

**Reply by tooptoop4** [reply]
i will try but is there any plan for single command? ie without needing to write out cr.csv

**Reply by jqnatividad** [reply]
`luau` should allow you to pull everything together in one script.

You can lookup https://github.com/dathere/qsv-lookup-tables to see how to pull together a luau script.

You can also check https://github.com/lancelijade/qqwry.lua/blob/master/crc32.lua for a pure Lua crc32 implementation.

Objectively speaking however, this is more "complex" than your already working implementation. I would just tweak it to skip the unnecessary `select` step and to use a `.csv` extension.  

Using `.csv` will allow `sqlp` to use polars' super-fast mem-mapped `read_csv` table function and should be dramatically faster than your original implementation that used a `.txt` extension.

```
/clis/qsv apply operations crc32 a_column --new-column bobb input.csv -o /tmp/output.csv
/clis/qsv sqlp /tmp/output.csv 'select sum(bobb) from _t_1'
```

**Reply by tooptoop4** [reply]
this is much slower than python's binascii.crc32 :(

**Reply by jqnatividad** [reply]
Do you have metrics you can share? How much slower? Can you share the python script you used using binascii.crc32?

Using the million row, 41 column, 520 MB sample of NYC's 311 data I use in qsv's benchmarks.

https://raw.githubusercontent.com/wiki/jqnatividad/qsv/files/NYC_311_SR_2010-2020-sample-1M.7z

it takes __2.384__ seconds on my 2023 Mac Mini with an M2 Pro, 12 CPU cores & 32GB of RAM:

```
$ hyperfine 'qsv apply operations crc32 "Unique Key" --new-column bobb NYC_311_SR_2010-2020-sample-1M.csv -o /tmp/output.csv; qsv sqlp /tmp/output.csv  "select sum(bobb) from _t_1"'
Benchmark 1: qsv apply operations crc32 "Unique Key" --new-column bobb NYC_311_SR_2010-2020-sample-1M.csv -o /tmp/output.csv; qsv sqlp /tmp/output.csv  "select sum(bobb) from _t_1"
  Time (mean ± σ):      2.384 s ±  0.011 s    [User: 2.879 s, System: 0.339 s]
  Range (min … max):    2.365 s …  2.399 s    10 runs
```

**Reply by jqnatividad** [reply]
I asked ChatGPT to write a python script using binascii.crc32:

```python
import pandas as pd
import binascii

def compute_crc32_and_sum(input_file, input_column, output_column, output_file):
    df = pd.read_csv(input_file)

    df[output_column] = df[input_column].apply(lambda x: binascii.crc32(str(x).encode('utf-8')) & 0xffffffff)

    checksum_sum = df[output_column].sum()

    df.to_csv(output_file, index=False)

    return checksum_sum

input_file = 'NYC_311_SR_2010-2020-sample-1M.csv'  # Input CSV file path
input_column = 'Unique Key'
output_column = 'crc32_column'
output_file = 'output.csv'

crc32_sum = compute_crc32_and_sum(input_file, input_column, output_column, output_file)
print(f"Sum of the new CRC32 column: {crc32_sum}")
```

and they both produced the same result:
```bash
$ qsv apply operations crc32 "Unique Key" --new-column bobb NYC_311_SR_2010-2020-sample-1M.csv -o /tmp/output.csv; qsv sqlp /tmp/output.csv  "select sum(bobb) from _t_1"          
bobb
2147188281889660
(1, 1)

$ python binascii-crc32.py            
Sum of the new CRC32 column: 2147188281889660
```
It takes _13.412_ seconds, 5.6x slower than the qsv equivalent above.

```
$ hyperfine 'python binascii-crc32.py'
Benchmark 1: python binascii-crc32.py
  Time (mean ± σ):     13.412 s ±  0.425 s    [User: 12.320 s, System: 0.975 s]
  Range (min … max):   13.036 s … 14.172 s    10 runs
```
So can you share your python script and timings? It's impossible to troubleshoot your problem with qsv being "much slower" otherwise, as my timings shows qsv is 5.6x faster.

**Reply by tooptoop4** [reply]
qsv apply operations crc32 "Unique Key" --new-column myfile.csv -o /tmp/output.csv along is taking 76seconds for my 6GB file with 203 columns (comma delimiter with each col enclosed in quotes)

**Reply by jqnatividad** [reply]
And how long did your python script take? If you're not willing to share your script, is the script above materially equivalent?

How many rows is your 6gb, 203 column file? Originally, you said you had 800k rows and 150 columns, so this must be another file.

The throughput in my example is 419,463.08 rows/seconds (1,000,000 rows/2.384 seconds). And that is because qsv apply is multi-threaded and the crc32fast library is SIMD-accelerated if your platform is supported.

In comparison, the throughput for my binascii-crc32.py is 74,560.10 rows/second (1,000,000 rows/13.412 seconds).

I suspect your docker configuration is the constraining factor that's preventing qsv from running efficiently.

Have you tried running `qsv apply operations crc32` outside your docker container?

Can you give the exact result of `qsv --version` inside your docker container and outside your docker container?

It will give you the number of threads qsv detected, and the memory available to it, which I'll require to help you troubleshoot your issue.

If you don't want to share diagnostic info I've been asking for since the beginning, then I'm afraid I can't help you any further.

**Reply by tooptoop4** [reply]
will be able to dig it up on monday, it took around 32 seconds in python, there was no pandas, it was a one liner along the lines of qsv select col numbers.txt | tail -n +2 | python -c "import sys, binascii; print(sum(binascii.crc32(line.encode()) for line in sys.stdin))"

**Reply by jqnatividad** [reply, marked as answer]
FWIW, with qsv 0.137.0, I get these figures:

```
$  hyperfine 'qsv apply operations crc32 "Unique Key" --new-column bobb NYC_311_SR_2010-2020-sample-1M.csv |  qsv sqlp - --cache-schema  "select sum(bobb) from _t_1"'
Benchmark 1: qsv apply operations crc32 "Unique Key" --new-column bobb NYC_311_SR_2010-2020-sample-1M.csv |  qsv sqlp - --cache-schema  "select sum(bobb) from _t_1"
  Time (mean ± σ):      2.802 s ±  0.008 s    [User: 3.076 s, System: 0.703 s]
  Range (min … max):    2.787 s …  2.813 s    10 runs
```

For your one-liner:
```
$ hyperfine 'qsv select 1 NYC_311_SR_2010-2020-sample-1M.csv | tail -n +2 | python -c "import sys, binascii; print(sum(binascii.crc32(line.encode()) for line in sys.stdin))"'
Benchmark 1: qsv select 1 NYC_311_SR_2010-2020-sample-1M.csv | tail -n +2 | python -c "import sys, binascii; print(sum(binascii.crc32(line.encode()) for line in sys.stdin))"
  Time (mean ± σ):     775.7 ms ±   3.3 ms    [User: 896.5 ms, System: 59.2 ms]
  Range (min … max):   769.6 ms … 779.3 ms    10 runs
```

So I recommend you stay with binascii, and just use qsv for selecting the column.
